### ISSUE TITLE
Warn when let shadows an outer let

### ISSUE THREAD

**Issue by danquirk** [palantir/tslint#497]
This just bit me replacing vars with lets in the TypeScript codebase. Existing code of the form:
```ts
function foo() {
    var text = "";
    if(something) {
          markerId = "stuff";
    } else {
          var markerId = "otherStuff";
    }
}
```
yields no errors from tsc or tslint. If you replace these vars with lets then you still get no errors, but you've changed the behavior of your code. Fortunately for us this was in relatively easy to diagnose text baselines (where the output is based on ```text```) but I can imagine much more difficult to debug cases.

**Comment by adidahiya** [comment]
That code reports a tslint failure for me when I turn on the `no-use-before-declare` rule, which it seems like you're also using ([here](https://github.com/Microsoft/TypeScript/blob/master/tslint.json))?

The `no-shadowed-variable` rule should check for things like this:

```ts
function foo() {
    let markerId = "";
    if (true) {
          markerId = "stuff";
    } else {
          let markerId = "otherStuff";
    }
}
```
Let me know if you run into any issues with it.

**Comment by adidahiya** [comment]
Hmm, now I'm not actually sure if `no-use-before-declare` should report a failure when the variable `markerId` has _technically_ already been declared via hoisting in code like that. It's probably good to be defensive against implicit hoisting, though, right?

Also, I _do_ get a compiler error when switching those vars to lets in your code @danquirk:
```
test.ts(5,11): error TS2304: Cannot find name 'markerId'.
```

**Comment by danquirk** [comment]
Yeah I think I had some settings in a weird state, seems correct now
